numpy/_core/src/umath/umathmodule.c

```diff
@@ -144,15 +145,75 @@ ufunc_frompyfunc(PyObject *NPY_UNUSED(dummy), PyObject *args, PyObject *kwds) {
     /* Do a better job someday */
     doc = "dynamic ufunc based on a python function";
 
+    /* Create the ufunc but without any loops by not passing types*/
     self = (PyUFuncObject *)PyUFunc_FromFuncAndDataAndSignatureAndIdentity(
             (PyUFuncGenericFunction *)pyfunc_functions, data,
-            types, /* ntypes */ 1, nin, nout, identity ? PyUFunc_IdentityValue : PyUFunc_None,
+            NULL, /* ntypes */ 0, nin, nout, identity ? PyUFunc_IdentityValue : PyUFunc_None,
             str, doc, /* unused */ 0, NULL, identity);
 
     if (self == NULL) {
         PyArray_free(ptr);
         return NULL;
     }
+
+    PyArray_DTypeMeta *op_dtypes[NPY_MAXARGS];
+    for (int arg = 0; arg < nin + nout; arg++) {
+        op_dtypes[arg] = PyArray_DTypeFromTypeNum(types[arg]);
+        /* These DTypes are immortal and adding INCREFs: so borrow it */
+        Py_DECREF(op_dtypes[arg]);
+    }
+
+    char method_name[101];
+    snprintf(method_name, 100, "legacy_ufunc_wrapper_for_%s", str);
+
+    NPY_ARRAYMETHOD_FLAGS flags = NPY_METH_REQUIRES_PYAPI | NPY_METH_NO_FLOATINGPOINT_ERRORS;
+
+    PyArrayMethod_GetReductionInitial *get_reduction_intial = NULL;
+    if (nin == 2 && nout == 1) {
+        npy_bool reorderable = NPY_FALSE;
+        PyObject *identity_obj = PyUFunc_GetDefaultIdentity(
+                self, &reorderable);
+        if (identity_obj == NULL) {
+            PyArray_free(ptr);
+            return NULL;
+        }
+        /*
+         * TODO: For object, "reorderable" is needed(?), because otherwise
+         *       we disable multi-axis reductions `arr.sum(0, 1)`. But for
+         *       `arr = array([["a", "b"], ["c", "d"]], dtype="object")`
+         *       it isn't actually reorderable (order changes result).
+         */
+        if (reorderable) {
+            flags |= NPY_METH_IS_REORDERABLE;
+        }
+        if (identity_obj != Py_None) {
+            get_reduction_intial = &get_initial_from_ufunc;
+        }
+    }
```

Comment:
If we use this API, we should just implement the `get_reduction_initial` directly I think?

Reply:
In that case, I suppose I will need to actually learn how get_reduction_initial should work. The only real example I see in the code base is the same get_initial_from_ufunc used here. Would we need something distinct here?

Reply:
Yeah, OK, so the main thing is that we need to store some additional data.  We can do that using:
```
_NPY_METH_static_data
```
That will contain the function and initial pointers (just FYI, right now they are leaked if the function would ever be cleaned up, but OK).

You can then retrieve those values via `context->method->static_data` (used in the `string_ufuncs.cpp`).

With that, you can just copy the `get_initial_from_ufunc`, except you always have an object and we can fetch the `identity` from `context->method->static_data` rather than the ufunc itself.